dask/array/gufunc.py

```diff
@@ -171,26 +202,108 @@ def apply_gufunc(func, signature, *args, **kwargs):
     if output_sizes is None:
         output_sizes = {}
 
+    ## Axes
+    ### Input validation
+    if (axes is not None) and (axis is not None):
+        raise ValueError("Only one of `axis` or `axes` keyword arguments should be given")
+    if axes and not isinstance(axes, list):
+        raise ValueError("`axes` has to be of type list")
+
+    _core_output_dimss = core_output_dimss if nout else [core_output_dimss]
+    _filtered_core_dims = list(filter(len, core_input_dimss))
+    _nr_outputs_with_coredims = len([True for x in _core_output_dimss if len(x) > 0])
+
+    if keepdims is True:
+        if _nr_outputs_with_coredims > 0:
+            raise ValueError("`keepdims` can only be used for scalar outputs")
+        _core_output_dimss = len(_core_output_dimss) * [_filtered_core_dims[0]]
+
+    _core_dims = core_input_dimss + _core_output_dimss
+    if axis is not None:
+        if not isinstance(axis, int):
+            raise ValueError("`axis` argument has to be an integer value")
+        if _filtered_core_dims:
+            cd0 = _filtered_core_dims[0]
+            if len(cd0) != 1:
+                raise ValueError("`axis` can be used only, if one core dimension is present")
+            for cd in _filtered_core_dims:
+                if cd0 != cd:
+                    raise ValueError("To use `axis`, all core dimensions have to be equal")
+
+    ### Expand dafaults or axis
+    if axes is None:
+        if axis is not None:
+            axes = [(axis,) if cd else tuple() for cd in _core_dims]
+        else:
+            axes = [tuple(range(-len(cid), 0)) for cid in _core_dims]
+    elif not isinstance(axes, list):
+        raise ValueError("`axes` argument has to be a list")
+    axes = [(a,) if isinstance(a, int) else a for a in axes]
+
+    ### Treat outputs
+    _nin = len(args)
+    _nout = nout if nout else 1
+
+    if ((_nr_outputs_with_coredims == 0) and (_nin != len(axes)) and (_nin + _nout != len(axes))) \
```

Comment:
nit: per PEP8, prefer adding additional parentheses to explicit continuations with `\`